In [30]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

### Import the all_goalkeepers_filtered.csv

(from Resources/Ben Reed)

In [31]:
# file directory
file_dir = "C:/Users/mamoo/UT_DataScience/UTA-VIRT-DATA-PT-02-2021-U-B/FinalProject/BenReed/"

# import the all_player.csv
all_gk_df = pd.read_csv(f"{file_dir}all_goalkeepers_filtered.csv")
pd.set_option("display.max_columns", None) # To show all columns explicitly
all_gk_df

Player Club POS  GP  GS  SHTS   SV  GA   W   L  T  ShO    W%  \
0            Mark Dodd  DAL  GK  31  31   235  160  45  17  14  0    6  54.8   
1           Tony Meola  MET  GK  29  29   188  142  38  14  15  0    9  48.3   
2       Mark Dougherty  TB   GK  28  28   186  123  47  17  11  0    3  60.7   
3         Jorge Campos  LA   GK  24  24   133  100  27  13   8  0    4  54.2   
4       Garth Lagerwey  KC   GK  23  22   133   83  38  12  10  0    3  52.2   
...                ...  ...  ..  ..  ..   ...  ...  ..  ..  .. ..  ...   ...   
2001   Adrian Zendejas  NaN  GK   0   0     0    0   0   0   0  0    0   0.0   
2002   Mason Stajduhar  NaN  GK   0   0     0    0   0   0   0  0    0   0.0   
2003       Brady Scott  NaN  GK   0   0     0    0   0   0   0  0    0   0.0   
2004  Kendall McIntosh  NaN  GK   0   0     0    0   0   0   0  0    0   0.0   
2005      Matt Bersano  NaN  GK   0   0     0    0   0   0   0  0    0   0.0   

       Sv%  Year Season  
0     68.1  1996    reg  
1     75.5  1996    reg  
2     66.1  1996    reg  
3     75.2  1996    reg  
4     62.4  1996    reg  
...    ...   ...    ...  
2001   0.0  2020   post  
2002   0.0  2020   post  
2003   0.0  2020   post  
2004   0.0  2020   post  
2005   0.0  2020   post  

[2006 rows x 16 columns]

In [32]:
# import the all_player.csv
all_gk1_df = pd.read_csv(f"{file_dir}all_goalkeepers_clean.csv")
pd.set_option("display.max_columns", None) # To show all columns explicitly
all_gk1_df

Player Club POS  GP  GS  SHTS   SV  GA   W   L  T  ShO    W%  \
0            Mark Dodd  DAL  GK  31  31   235  160  45  17  14  0    6  54.8   
1           Tony Meola  MET  GK  29  29   188  142  38  14  15  0    9  48.3   
2       Mark Dougherty  TB   GK  28  28   186  123  47  17  11  0    3  60.7   
3         Jorge Campos  LA   GK  24  24   133  100  27  13   8  0    4  54.2   
4       Garth Lagerwey  KC   GK  23  22   133   83  38  12  10  0    3  52.2   
...                ...  ...  ..  ..  ..   ...  ...  ..  ..  .. ..  ...   ...   
2001   Adrian Zendejas  NaN  GK   0   0     0    0   0   0   0  0    0   0.0   
2002   Mason Stajduhar  NaN  GK   0   0     0    0   0   0   0  0    0   0.0   
2003       Brady Scott  NaN  GK   0   0     0    0   0   0   0  0    0   0.0   
2004  Kendall McIntosh  NaN  GK   0   0     0    0   0   0   0  0    0   0.0   
2005      Matt Bersano  NaN  GK   0   0     0    0   0   0   0  0    0   0.0   

       Sv%  Year Season  
0     68.1  1996    reg  
1     75.5  1996    reg  
2     66.1  1996    reg  
3     75.2  1996    reg  
4     62.4  1996    reg  
...    ...   ...    ...  
2001   0.0  2020   post  
2002   0.0  2020   post  
2003   0.0  2020   post  
2004   0.0  2020   post  
2005   0.0  2020   post  

[2006 rows x 16 columns]

Both the **all_goalkeepers_filtered.csv** and **all_goalkeepers_clean.csv** are the same files created by *Ben Reed*.

In [33]:
# Check all the column headers.
all_gk_df.columns

Index(['Player', 'Club', 'POS', 'GP', 'GS', 'SHTS', 'SV', 'GA', 'W', 'L', 'T',
       'ShO', 'W%', 'Sv%', 'Year', 'Season'],
      dtype='object')

Most of the column headers are in abbreviated form which is hard to understand. We need to rename columns with a little elaborate titles.

In [34]:
all_gk_df = all_gk_df.rename(columns={"POS": "Position", 
                               "GP": "GamesPlayed",
                              "GS": "GamesStarted",
                              "SHTS": "ShotsFaced",
                              "SV": "SavesMade",
                              "GA": "GoalsAgainst",
                              "W": "Wins",
                              "L": "Losses",
                              "T": "Ties",
                              "ShO": "ShutOuts",
                              "W%": "WinsPercentage",
                              "Sv%": "SavePercentage"
                              })
all_gk_df.head(2)

Player Club Position  GamesPlayed  GamesStarted  ShotsFaced  SavesMade  \
0   Mark Dodd  DAL       GK           31            31         235        160   
1  Tony Meola  MET       GK           29            29         188        142   

   GoalsAgainst  Wins  Losses  Ties  ShutOuts  WinsPercentage  SavePercentage  \
0            45    17      14     0         6            54.8            68.1   
1            38    14      15     0         9            48.3            75.5   

   Year Season  
0  1996    reg  
1  1996    reg

In [35]:
# Show any rows where 'Club' title is empty
all_gk_df[(all_gk_df['Club'] == '') | (all_gk_df['Club'].isnull())]

Player Club Position  GamesPlayed  GamesStarted  ShotsFaced  \
1231      Mark Simpson  NaN       GK            6             6          28   
1232      Jorge Campos  NaN       GK            6             6          35   
1233    Garth Lagerwey  NaN       GK            5             5          31   
1234    Mark Dougherty  NaN       GK            4             4          17   
1235      David Kramer  NaN       GK            3             0           3   
...                ...  ...      ...          ...           ...         ...   
2001   Adrian Zendejas  NaN       GK            0             0           0   
2002   Mason Stajduhar  NaN       GK            0             0           0   
2003       Brady Scott  NaN       GK            0             0           0   
2004  Kendall McIntosh  NaN       GK            0             0           0   
2005      Matt Bersano  NaN       GK            0             0           0   

      SavesMade  GoalsAgainst  Wins  Losses  Ties  ShutOuts  WinsPercentage  \
1231         21             7     5       0     0         1            83.3   
1232         25             6     2       2     0         1            33.3   
1233         18             9     2       3     0         0            40.0   
1234         10             7     2       2     0         1            50.0   
1235          2             0     2       0     0         1            66.7   
...         ...           ...   ...     ...   ...       ...             ...   
2001          0             0     0       0     0         0             0.0   
2002          0             0     0       0     0         0             0.0   
2003          0             0     0       0     0         0             0.0   
2004          0             0     0       0     0         0             0.0   
2005          0             0     0       0     0         0             0.0   

      SavePercentage  Year Season  
1231            75.0  1996   post  
1232            71.4  1996   post  
1233            58.1  1996   post  
1234            58.8  1996   post  
1235            66.7  1996   post  
...              ...   ...    ...  
2001             0.0  2020   post  
2002             0.0  2020   post  
2003             0.0  2020   post  
2004             0.0  2020   post  
2005             0.0  2020   post  

[682 rows x 16 columns]

So there are **682 rows** out of 2006 rows where the *Club* column has no value. In our analysis, the "Club" information is crucial. So we may drop the rows where there is no info for *Club*. 

In [36]:
all_gk_df = all_gk_df.drop(all_gk_df[all_gk_df['Club'].isnull()].index)
all_gk_df

Player Club Position  GamesPlayed  GamesStarted  ShotsFaced  \
0            Mark Dodd  DAL       GK           31            31         235   
1           Tony Meola  MET       GK           29            29         188   
2       Mark Dougherty  TB        GK           28            28         186   
3         Jorge Campos  LA        GK           24            24         133   
4       Garth Lagerwey  KC        GK           23            22         133   
...                ...  ...      ...          ...           ...         ...   
1940    Emmanuel Ochoa   SJ       GK            0             0           0   
1941   Drake Callender  MIA       GK            0             0           0   
1942    Elliot Panicco  NSH       GK            0             0           0   
1943         Eric Dick  SKC       GK            0             0           0   
1944  Stefan Cleveland  SEA       GK            0             0           0   

      SavesMade  GoalsAgainst  Wins  Losses  Ties  ShutOuts  WinsPercentage  \
0           160            45    17      14     0         6            54.8   
1           142            38    14      15     0         9            48.3   
2           123            47    17      11     0         3            60.7   
3           100            27    13       8     0         4            54.2   
4            83            38    12      10     0         3            52.2   
...         ...           ...   ...     ...   ...       ...             ...   
1940          0             0     0       0     0         0             0.0   
1941          0             0     0       0     0         0             0.0   
1942          0             0     0       0     0         0             0.0   
1943          0             0     0       0     0         0             0.0   
1944          0             0     0       0     0         0             0.0   

      SavePercentage  Year Season  
0               68.1  1996    reg  
1               75.5  1996    reg  
2               66.1  1996    reg  
3               75.2  1996    reg  
4               62.4  1996    reg  
...              ...   ...    ...  
1940             0.0  2020    reg  
1941             0.0  2020    reg  
1942             0.0  2020    reg  
1943             0.0  2020    reg  
1944             0.0  2020    reg  

[1324 rows x 16 columns]

In [37]:
# Find the total number of unique Club names.
all_gk_df['Club'].nunique()

41

In [38]:
# Find the unique Club name abbreviations.
all_gk_df['Club'].unique()

array(['DAL', 'MET', 'TB ', 'LA ', 'KC ', 'COL', 'SJ ', 'NE ', 'DC ',
       'CLB', 'CHI', 'MIA', 'RSL', 'ATL', 'CHV', 'HOU', 'NY ', 'TOR',
       'DC', 'NE', 'SEA', 'NYC', 'SKC', 'POR', 'PHI', 'VAN', 'NSH', 'LA',
       'MTL', 'RBNY', 'ORL', 'CAN', 'BRA', 'ALG', 'NYR', 'MIN', 'NY',
       'CIN', 'SJ', 'LFC', 'LAFC'], dtype=object)

In [39]:
# Rename the 'Club' column's club abbreviations.
all_gk_df = all_gk_df.replace({'Club': {'DAL':'FC Dallas',
                                        'MET':'Metropolitan Oval',
                                        'TB':'Tampa Bay United',
                                        'LA ':'LA Galaxy',
                                        'KC':'Sporting Kansas City',
                                        'COL':'Colorado Rapids',
                                        'SJ ':'San Jose Earthquakes',
                                        'NE ':'New England Revolution',
                                        'DC ':'D.C. United',
                                        'CLB':'Columbus Crew', 
                                        'CHI':'Chicago FIre FC',
                                        'MIA':'Inter Miami CF',
                                        'RSL':'Real Salt Lake',
                                        'ATL':'Atlanta United',
                                        'CHV':'Chivas USA',
                                        'HOU':'Houston Dynamo FC',
                                        'NY ': 'New York City',
                                        'TOR':'Toronto FC',
                                        'DC':'D.C. United',
                                        'NE':'New England Revolution',
                                        'SEA':'Seattle Sounders FC',
                                        'NYC':'New York City',
                                        'SKC':'Sporting Kansas City',
                                        'POR':'Portland Timbers',
                                        'PHI':'Philadelphia Union',
                                        'VAN':'Vancouver Whitecaps FC',
                                        'NSH':'Nashville SC',
                                        'LA':'LA Galaxy',
                                        'MTL':'CF Montreal',
                                        'RBNY':'New York City Red Bulls',
                                        'ORL':'Orlando City SC',
                                        'NY': 'New York City',
                                        'MIN':'Minnesota United FC',
                                        'CIN':'FC Cincinnati',
                                        'SJ':'San Jose Earthquakes',
                                        'LFC':'Los Angeles FC', 
                                        'LAFC':'Los Angeles FC', 
                                        }})
all_gk_df.tail()

Player                  Club Position  GamesPlayed  \
1940    Emmanuel Ochoa  San Jose Earthquakes       GK            0   
1941   Drake Callender        Inter Miami CF       GK            0   
1942    Elliot Panicco          Nashville SC       GK            0   
1943         Eric Dick  Sporting Kansas City       GK            0   
1944  Stefan Cleveland   Seattle Sounders FC       GK            0   

      GamesStarted  ShotsFaced  SavesMade  GoalsAgainst  Wins  Losses  Ties  \
1940             0           0          0             0     0       0     0   
1941             0           0          0             0     0       0     0   
1942             0           0          0             0     0       0     0   
1943             0           0          0             0     0       0     0   
1944             0           0          0             0     0       0     0   

      ShutOuts  WinsPercentage  SavePercentage  Year Season  
1940         0             0.0             0.0  2020    reg  
1941         0             0.0             0.0  2020    reg  
1942         0             0.0             0.0  2020    reg  
1943         0             0.0             0.0  2020    reg  
1944         0             0.0             0.0  2020    reg

In [40]:
# Check if there is still any Null value.
all_gk_df.isnull().sum()

Player            0
Club              0
Position          0
GamesPlayed       0
GamesStarted      0
ShotsFaced        0
SavesMade         0
GoalsAgainst      0
Wins              0
Losses            0
Ties              0
ShutOuts          0
WinsPercentage    0
SavePercentage    0
Year              0
Season            0
dtype: int64

It looks like there is **no null values** (in total of 1324 rows).

In [41]:
# Drop the column "xxx"
#all_gk_df = all_players_df.drop("xxx", axis='columns')
#all_gk_df.head(2)

In [42]:
# The "yyy" column has also some inconsistent data. We're focused on the team or club.
# As we still have 28 columns of data, we decide to to drop this column.
#all_gk_df = all_gk_df.drop("yyy", axis='columns')
#all_gk_df.head(2)

In [43]:
all_gk_df.describe()

GamesPlayed  GamesStarted   ShotsFaced    SavesMade  GoalsAgainst  \
count  1324.000000   1324.000000  1324.000000  1324.000000   1324.000000   
mean      9.737915      9.558912    48.539275    33.313444     13.802115   
std      11.597432     11.603828    59.008070    41.363833     16.485504   
min       0.000000      0.000000     0.000000     0.000000      0.000000   
25%       0.000000      0.000000     0.000000     0.000000      0.000000   
50%       4.000000      3.000000    17.000000    11.000000      6.000000   
75%      20.000000     19.000000    93.250000    64.000000     27.000000   
max      34.000000     34.000000   249.000000   184.000000     64.000000   

              Wins       Losses         Ties     ShutOuts  WinsPercentage  \
count  1324.000000  1324.000000  1324.000000  1324.000000     1324.000000   
mean      3.743202     3.742447     2.071752     2.277946       22.063293   
std       5.062026     4.629686     3.186112     3.281884       23.986060   
min       0.000000     0.000000     0.000000     0.000000        0.000000   
25%       0.000000     0.000000     0.000000     0.000000        0.000000   
50%       1.000000     1.000000     0.000000     0.000000       20.000000   
75%       7.000000     7.000000     4.000000     4.000000       40.000000   
max      22.000000    19.000000    18.000000    16.000000      100.000000   

       SavePercentage         Year  
count     1324.000000  1324.000000  
mean        43.322885  2009.228097  
std         33.238269     8.439600  
min          0.000000  1996.000000  
25%          0.000000  2004.000000  
50%         61.550000  2011.000000  
75%         69.700000  2016.000000  
max        100.000000  2020.000000

In [20]:
all_gk_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1324 entries, 0 to 1944
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Player          1324 non-null   object 
 1   Club            1324 non-null   object 
 2   Position        1324 non-null   object 
 3   GamesPlayed     1324 non-null   int64  
 4   GamesStarted    1324 non-null   int64  
 5   ShotsFaced      1324 non-null   int64  
 6   SavesMade       1324 non-null   int64  
 7   GoalsAgainst    1324 non-null   int64  
 8   Wins            1324 non-null   int64  
 9   Losses          1324 non-null   int64  
 10  Ties            1324 non-null   int64  
 11  ShutOuts        1324 non-null   int64  
 12  WinsPercentage  1324 non-null   float64
 13  SavePercentage  1324 non-null   float64
 14  Year            1324 non-null   int64  
 15  Season          1324 non-null   object 
dtypes: float64(2), int64(10), object(4)
memory usage: 175.8+ KB


The **all_gk_df** is now **cleaned** with total of **16 columns, 1344 rows** of non-null data with consistent data types. 

### Export the clean df to .csv

In [23]:
# Drop the 'Season' and 'Position' column as it doesn't have any significant input to the analysis.
all_gk_cleaned = all_gk_df.drop(["Season", "Position"], axis='columns')
all_gk_cleaned.head(2)

Player               Club  GamesPlayed  GamesStarted  ShotsFaced  \
0   Mark Dodd          FC Dallas           31            31         235   
1  Tony Meola  Metropolitan Oval           29            29         188   

   SavesMade  GoalsAgainst  Wins  Losses  Ties  ShutOuts  WinsPercentage  \
0        160            45    17      14     0         6            54.8   
1        142            38    14      15     0         9            48.3   

   SavePercentage  Year  
0            68.1  1996  
1            75.5  1996

In [28]:
# Export the clean dataframe to "all_gk_cleaned.csv"
csv_dir = "C:/Users/mamoo/UT_DataScience/UTA-VIRT-DATA-PT-02-2021-U-B/FinalProject/Resources/Clean_csv"
all_gk_cleaned.to_csv(f"{csv_dir}all_gk_cleaned.csv")

In [25]:
all_gk_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1324 entries, 0 to 1944
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Player          1324 non-null   object 
 1   Club            1324 non-null   object 
 2   GamesPlayed     1324 non-null   int64  
 3   GamesStarted    1324 non-null   int64  
 4   ShotsFaced      1324 non-null   int64  
 5   SavesMade       1324 non-null   int64  
 6   GoalsAgainst    1324 non-null   int64  
 7   Wins            1324 non-null   int64  
 8   Losses          1324 non-null   int64  
 9   Ties            1324 non-null   int64  
 10  ShutOuts        1324 non-null   int64  
 11  WinsPercentage  1324 non-null   float64
 12  SavePercentage  1324 non-null   float64
 13  Year            1324 non-null   int64  
dtypes: float64(2), int64(10), object(2)
memory usage: 155.2+ KB


In [26]:
# Group by "Club" and "Year"
all_gk_grouped = all_gk_cleaned.groupby(['Club','Year']).sum()
all_gk_grouped

GamesPlayed  GamesStarted  ShotsFaced  SavesMade  \
Club                   Year                                                     
ALG                    2014            4             4          13          7   
                       2015            5             5          20         10   
Atlanta United         2005           24            23         156        104   
                       2006           13            13          70         49   
                       2007           27            27         119         87   
...                                  ...           ...         ...        ...   
Vancouver Whitecaps FC 2016           34            34         172        116   
                       2017           37            36         185        125   
                       2018           25            25         118         68   
                       2019           26            26         159        114   
                       2020           24            23         131         89   

                             GoalsAgainst  Wins  Losses  Ties  ShutOuts  \
Club                   Year                                               
ALG                    2014             4     1       1     2         1   
                       2015             9     0       3     2         1   
Atlanta United         2005            46     3      16     4         2   
                       2006            19     3       6     4         5   
                       2007            25    14       6     7        13   
...                                   ...   ...     ...   ...       ...   
Vancouver Whitecaps FC 2016            52    10      15     9         7   
                       2017            55    15      14     7         7   
                       2018            48    10      11     4         2   
                       2019            46     7      13     6         5   
                       2020            44     9      14     0         3   

                             WinsPercentage  SavePercentage  
Club                   Year                                  
ALG                    2014            25.0            53.8  
                       2015             0.0            50.0  
Atlanta United         2005            12.5            66.7  
                       2006            23.1            70.0  
                       2007            51.9            73.1  
...                                     ...             ...  
Vancouver Whitecaps FC 2016            30.3           166.3  
                       2017            86.4           255.7  
                       2018            41.7           108.0  
                       2019            26.9            71.7  
                       2020           119.4           267.4  

[358 rows x 11 columns]

In [29]:
# Export the grouped dataframe to "all_gk_grouped.csv"
csv_dir = "C:/Users/mamoo/UT_DataScience/UTA-VIRT-DATA-PT-02-2021-U-B/FinalProject/Resources/Clean_csv"
all_gk_grouped.to_csv(f"{csv_dir}all_gk_grouped.csv")